src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -1666,52 +1666,101 @@ IEnumerable<List<ForeignKeyDefinition>> foreignKeysForAllTargetEntities
-                                // specified in the configuration file.
-                                if (fk.ReferencedColumns.Count == 0)
-                                {
-                                    fk.ReferencedColumns.AddRange(inferredDefinition.ReferencedColumns);
-                                }
-                            }
-                        }
+                // Add the referencing and referenced columns for this foreign key definition for the target.
+                if (PairToFkDefinition is not null &&
+                    PairToFkDefinition.TryGetValue(fKDefinitionToTarget.Pair, out ForeignKeyDefinition? inferredFKDefinition))
+                {
+                    // Being here indicates that we inferred an FK constraint for the current FK.
+                    // The count of referencing and referenced columns being > 0 indicates that source.fields and target.fields 
+                    // have been specified in the config file. In this scenario, higher precedence is given to the fields configured through the config file. So, the existing FK definition is retained as is.   
+                    if (fKDefinitionToTarget.ReferencingColumns.Count > 0 && fKDefinitionToTarget.ReferencedColumns.Count > 0)
+                    {
+                        validatedFKDefinitionsToTarget.Add(fKDefinitionToTarget);
+                    }
+                    // The count of referenced and referencing columns being = 0 indicates that source.fields and target.fields 
+                    // are not configured through the config file. In this case, the FK fields inferred from the database are populated.
+                    else
+                    {
+                        validatedFKDefinitionsToTarget.Add(inferredFKDefinition);
+                    }
+                }
+                else
+                {
+                    // This code block adds FK definitions between source and target entities when there is no FK constraint defined
+                    // in the database, either from source->target or target->source entities.
+
+                    // Being here indicates that we did not find an FK constraint in the database for the current FK definition.
+                    // But this does not indicate absence of an FK constraint between the source, target entities yet.
+                    // This may happen when an FK constraint exists between two tables, but in an order opposite to the order
+                    // of referencing and referenced tables present in the current FK definition. This happens because for a relationship
+                    // with right cardinality as 1, we add FK definitons from both source->target and target->source to the source entity's definition.
+                    // because at that point we don't know if the relationship is an N:1 relationship or a 1:1 relationship.
+                    // So here, we need to remove the wrong FK definition for:
```

Comment:
One thing that tripped us up during review was the usage of 'wrong'. Is it accurate to say that 'wrong' can be replaced with either 'redundant' or 'duplicate' ? Because it looks like, without this check, the where clause had duplicate predicates, but wasn't exactly causing wrong results/crashing.

Reply:
If you talk about Foreign key, they are from referencing table to referencing table. So when you add a foreign key from publisher as referencing table to book as referenced table, this is a wrong foreign key and not redundant.

I get why you have the idea of redundant - because for joins it does not matter which is the referencing entity. But going forward for mulitple-mutation, this matters.